### DISCUSSION TITLE
Project Manager & Project List UX improvements

### DISCUSSION THREAD

**Discussion by mgdawber** [godotengine/godot-proposals#9121] [Editor]
I've noticed a few things which could be improved upon in the Project Manager & Project List, and I've given it a go with implementing them. Please let me know your thoughts!

**Update: The design draft has changed quite a bit since the original comment. I've hidden it away within this details section.**

<details><summary>First Draft</summary>

**Main Changes**

- [ ] 'Manage Tags' should have it's own icon - at the minute it's using the 'script' icon.
- [ ] Both the Tags & Icon buttons (Favorite/File Explorer) are not clear that they are clickable upon first glance, as they both use non-standard button styling. In this proposal, I've removed the custom styling, and moved the the tags closer to the project name and file path.
- [ ] Improve the hover aspect of the buttons so that it's less drastic when hovering over the icon buttons (It clashes with the project hover highlight) & some have inconsistent states.
- [ ] Reduce the font size of the main project title to be more consistent with the rest of the information displayed.
- [ ] Add a tag icon, and move the favorite icon so that it sits next to the project title, forming a vertical bar of buttons - it's easier to see that they're clickable.

**Vanity Changes**

These changes are definitely not needed and are a personal choice, but no harm in presenting them 😜.

- [ ] Change the Favorite Icon to a golden colour.
- [ ] Remove the Sort Label and give the sort option a minimum width to coincide with the project manager options. (This isn't great, I just like the way it lines up.)
- [ ] Remove the random colour attached to a tag. I really like the idea of this, but think it could be implemented in a manner where the user can assign colours to tags to add meaning. Give tags red for 'Abandoned', 'Needs Revision', or orange for 'Draft' etc.

Before & After:

<img width="1624" alt="image" src="https://github.com/godotengine/godot-proposals/assets/61522828/d616ef75-f0d1-4dc9-bafe-a4cebe3b337f">

<img width="1624" alt="image" src="https://github.com/godotengine/godot-proposals/assets/61522828/135e30e9-7828-41bd-bcbc-2d61a9f9bcde">

Before & After (Hover States):
 
https://github.com/godotengine/godot-proposals/assets/61522828/6fdb00f8-33f2-4072-981b-8e44b69a0c2e

https://github.com/godotengine/godot-proposals/assets/61522828/9569864f-b8c2-4479-a0c7-34c9bd27038b
</details> 

**Latest version as of 20 Feb 2024**

![image](https://github.com/godotengine/godot-proposals/assets/61522828/fc688a99-c532-4adc-aec7-c391f59e74df)

**Latest versions as of 26 Feb 2024**

![MacBook Air - 2 (1)](https://github.com/godotengine/godot-proposals/assets/61522828/0b7bc6af-2185-42a7-ace2-b34d46069d07)
![MacBook Air - 3 (1)](https://github.com/godotengine/godot-proposals/assets/61522828/2a6f8910-05b0-49aa-a50c-e64361e5fb4d)

Please let me know your thoughts, is there anything that you like/dislike? Thanks!

**Comment by Calinou** [comment]
While it looks good aesthetically, one issue I have with displaying tags below the project name and path is that it increases the height taken per project. People may have hundreds of projects in their list, so we need to fit a reasonable amount of projects per page even on small displays to avoid an excessive amount of scrolling.

There may be ways to optimize this (such as decreasing the spacing between name/path/tags or displaying tags with a smaller font size), but preserving screen real estate is important.

> Remove the random colour attached to a tag. I really like the idea of this, but think it could be implemented in a manner where the user can assign colours to tags to add meaning. Give tags red for 'Abandoned', 'Needs Revision', or orange for 'Draft' etc.

The reason tags have randomly assigned colors based on their name is that it allows standardizing the colors across projects. Otherwise, users would have to manually define tag colors in the *editor* settings (as opposed to project settings), which would result in tag colors being inconsistent across users if they are editing the same project. Defining tag colors in project settings would be problematic, as it would allow using different colors for the same tag in different projects (therefore nullifying the usefulness of colors to ease visual grepping).

While this approach takes away control over controls from the user, it's more reliable and deterministic for all users.

**Reply by mgdawber** [reply]
ah thank you for the response, that makes a lot of sense now! 

I've made another scenario with the adjustments in place, putting the tags back.

**Comment by passivestar** [comment]
Sorry for hijacking the discussion, but since it's about the project manager ui/ux in general I'll share some ideas here, maybe you'll find something useful:

1. Having a sidebar for operations is completely avoidable. Right now you need to click on the project and then click edit/run/rename/whatever, which is 2 clicks. You can put those into an ellipsis popup menu drawn on the right of each project entry and it will still be 2 clicks, but you'll be able to get rid of that sidebar. The popup menu can also be called via right click. Even better you can take the commonly used ones out of the popup menu (like remove) and make them icon-only buttons on entries, and leave rarely used (like rename) in the popup menu. That way removing becomes one-click and you can remove multiple projects quickly

2. The Edit button probably shouldn't exist when you can double click entries. If you remove the sidebar with its "click on entry then click on button" workflow you can even change it to a single click if you want

3. The folder icon is unnecessary when you can just make the path string itself clickable (it wasn't obvious to me that the folder icon is a button anyway, especially before this redesign)

4. Manage Tags popup probably shouldn't exist. When you hover over a tag it could show you a small cross icon in it to remove the tag. To the right of the list of tags there could be a small button with a "+" that lets you add a tag inline. Sort of what you'd usually see in web applications when editing content tags

5. The Filter Projects input field is too wide, there's no way you're typing that much in it. It should have a cap on max width. It's better to have whitespace, otherwise it won't look good on large monitors

6. The Create button is more important than all of the other buttons, it should look different, i.e use an accent color. It's the primary action you perform in that window unless you're opening an existing project

7. This is minor, but personally I'd change the Settings icons to a cog because it's more widely adopted for settings in other apps so it's easier to recognize

### Some feedback on your redesign:

1. I like that you made the tags less contrasty, they are supposed to be metadata and they were drawing too much attention. I'm not sure how I feel about the original design with the stripe on the left though, I would have probably experimented with some other options (i.e stipe on the bottom/top, solid colored background, corner radii, etc) because I'm having trouble identifying what the stripe belongs to at a glance, but this might be subjective

2. I like the colored stars. Maybe could try keeping the disabled ones monochrome though

**Reply by mgdawber** [reply]
Not at all - Great feedback, thank you very much!

I pretty much agree with all of proposed ideas - I'm halfway through creating a mock-up with these proposals so I'll post it here when I'm done.

1. Very much like this idea, I'll finish the mock-up with a potential for what it could look like.

3. Agreed. There should still be some visual aspect that it's clickable - A lot of applications change to an accent colour + underline on hover so maybe we could try that.

> I would have probably experimented with some other options (i.e stipe on the bottom/top, solid colored background, corner radii, etc) because I'm having trouble identifying what the stripe belongs to at a glance, but this might be subjective

I didn't add this to my original comment but I also thought this - glad to see it wasn't just me 😄

**Reply by mgdawber** [reply]
So I think I got everything, let me know what you think!

![redesign](https://github.com/godotengine/godot-proposals/assets/61522828/30485bcc-fac9-4ff6-a40a-f1a90e58fd01)


Changes from previous version:

- Create button now has accent colour background.
- Settings button icon changed to a cog wheel. I've added bold to align with the rest of the top bar, but may not be necessary.
- Filter and Sort by are now aligned centrally. 
- Removed side bar in favour of icon buttons, aligned bottom right.
- Changed tag colour to apply to the border and background. The background has an opacity of 15% so that there is no conflicting contrast with the white text.
- Added hover states to file path, play and delete as shown in the last project. (The standard hover may do for play/delete.. Thought I'd add this is just in case any one has a preference.)

I also changed the 'Import' Icon to another one that I found within the project - Potentially more fitting?

An added bonus of this is that 'Remove Missing' becomes part of the upper bar. As of now, it is contained within the actions relative to a single project.

_Edit: Looking at it again the tags might be a bit too colourful, can't decide._
_Edit 2: I forgot to add the +/x for tags 😵‍💫_

**Reply by passivestar** [reply]
Hey, I dragged some things around in your screenshot, let me know what you think

Things I changed:

1. Returned the star icon to the left (the design was a bit too right-heavy, especially with all of the icons and colors there). Also dimmed the inactive state of it

2. Split the searchbar and sorting on top (the top bar was starting to look like a dashed line with all those spaces)

3. Changed the thing on the right to be one-row

4. Not sure how often people run projects without editing them from the project manager? The run button might be a candidate for the sandwich menu

5. Swapped the godot icon for a solid one from the official press-kit, the outline version isn't working that well at that scale imo

6. Added some space around the whole thing

![gd](https://github.com/godotengine/godot-proposals/assets/60579014/2addd6a6-1fca-4c24-9de5-f50a4c5a01b1)

Notes:

1. Not a huge fan of the border around the whole thing tbh and the fact that the top buttons are visually INSIDE of that border, it creates some awkward situations like what we have to the left of the Create button in my version. It's possible that there's a version of this with no border around the entire thing, where the top bar is just a one long rectangle, and then areas below and above it use the same background color. This would need testing and iteration to figure out though

2. I do believe that solid-color tags are an improvement. Maybe the outlines could be a tiny bit less contrasty, but other than that they are cleaner than what we had before

3. Cog _is_ an improvement, looks much cleaner

4. The color of the Create button might need to be a little bit darker to make the text on it more readable

5. Forgot to add back the list separators in my version btw. Those could use low contrast and thicker lines (2px instead of 1px)

6. The import icon with an arrow is more fitting, it's a shame it's an outline icon with worse readability, so I'm a bit conflicted on this one

7. A bit of an off-topic but a window this wide is not the best for the presentation anymore with the sidebar removed, the default project manager window width will need to be adjusted to account for that too (for an average project name length and average number of tags). Couldn't change that in my version because was cutting your screenshot and that would just be too much cutting 😄

**Reply by mgdawber** [reply]
This is a big improvement imo, very clean!

I had a go at fixing the things you mentioned with the design.

![Wireframe - 1](https://github.com/godotengine/godot-proposals/assets/61522828/cd71ecf0-baf3-44c8-bf43-3459c300a0c4)


- Slightly dulled the create button.
- Removed the 'background border'. Agreed that other UI's could use this - we would have to see what the asset library looked liked without it.
- Had a go at implementing the import icon to be more in line with the others.
- Reduced opacity of tag borders to 80%.

_Edit: I may have misinterpreted which border to remove 😅?_

**Reply by passivestar** [reply]
OOOOO I love what you did, you didn't misinterpret anything

I just did a tiny pass on top of it to get this:

![gddes](https://github.com/godotengine/godot-proposals/assets/60579014/07697224-9519-4eea-8b3f-b42a0b6daeb1)

- Added a box around the "top bar" (had to just blend it on top for this sketch), imo it helps with giving the entire thing a little bit of structure without any unnecessary borders, but this is just a quick sketch and can be further iterated upon if we like this direction
- Added some more vertical space around the top bar
- Removed the separator from the last list item
- Added the missing icon to the search field
- Moved the paths a bit to the right

This is starting to look good imo, one thing that could be improved that's kind of noticeable is the lack of horizontal paddings inside of the buttons with text, but I must say that that is an issue in the entirety of Godot, so perhaps this should be addressed on the app-wide level

Also yeah the asset library will have to be redesigned for this look too, but I'm fairly confident that that is achievable

One thing that I must note is that this is somewhat a step away from the current Godot design language (the editor relies on borders heavily), but I think the launcher can have it's own unique feel in terms of whitespace and whatnot, a lot of launchers are like that, Unity and Unreal are the big examples of this. Also the editor's UI itself is still not mature, a lot of issues will have to be addressed, especially paddings and borders

P.S. Your import icon is awesome
P.P.S. Since those list items are now basically large buttons you'll need to design hover states for them. I'm thinking maybe a rounded box around the whole thing, that's only visible on hover (although if we do rounded corners on them we can't keep the separators)

**Reply by davthedev** [reply]
I am perhaps a bit old school and use my inspiration in good old desktop apps rather than websites.

Personally, I would like to keep the two-click approach and the side actions which are a standard pattern in desktop productivity apps because
- Actions buttons have a fixed place and are consistent
- You can select multiple items then do the action once. 
- Items can be more compact because the room taken by the embedded action buttons is freed up, thus you can see more of them at a glance.
- You can navigate the list by keyboard. One hit = next row. Instead, with embedded action buttons, there would need multiple hits to traverse all the buttons before heading to the next item, and you have to make sure your focus is on the actual action you need.
 
For example, if there is a delete operation to run on multiple projects, this would give one final opportunity to confirm rather than click... confirm one... click... confirm one... click... confirm one... click... confirm one... hypnosis... click... confirm one... -OOPS! regret was the wrong one- 
The mindset when selecting multiple is to prepare the batch you want to act on before you do the action.

Perhaps because I grew up with computers before the advent of everything done on web apps. I still would like to be able to do the same on some web apps, have the select-then-act approach rather than pinpoint the action on the item row itself. We can do a mix of both approaches.

If there is an action doable by double click, it should not be the only way to run the said action. There should be an equivalent to access it by a regular action button or menu item. I happen to quickly open some projects for demos on my touchscreen computer during meetups and I am grateful that all actions, including Open project, are available by buttons.

One issue with hyperlinks, if not handled properly, is that the hitbox size depends on the text length. Make sure the hitbox size remains consistent, for example by making it take the full available width and make an "area" hover/pressed effect that takes the entire width, like most buttons. A simple link in this context does not give a clue of what it does. As this is not a web browser, we have to clarify that it will actually open the external file browser. I would prefer an actual "Open in file browser" action. Or put the folder icon back so that nobody will confuse it with a web URL!

The big padding is not friendly for small screens IMHO. My project manager window does not open that big. There may be a need for a compact theme in this case. 
It makes more difficult figuring out where the boundaries of the scrollable list are located, because, in desktop apps, there is no assumption that the whole window contents is a giant scrollable area by default. Windows are fixed surfaces with eventual scrollable areas that are generally delimited visually, so that you know in which area you need to point your mouse for the scroll wheel to work.

**Reply by passivestar** [reply]
You'd generally want to optimize the UI for what you do the most

What do users do the most in that window? How often do users remove multiple projects compared to running a specific project? This is one consideration.

And another one is user-friendliness. This window is what every new user will see first when they download and run the engine. How do you want this window to feel? What do you want it to say? Do you want it to feel like creating a new project in Godot is easy and fun, or do you want it to feel like you just opened Total Commander and you better start reading the manual?

I know, it's a bit of a tradeoff between what the existing users are used to and what is needed to make the new users stick, but those are the question you need to ask yourself if you want the engine to become mainstream

> You can select multiple items then do the action once

I agree that we could do a mix of both approaches, and it's probably the best thing to do here

What we could do is when you click on an entry you don't run it, and if you select multiple entries by holding shift an additional row of buttons appears below the top bar for batch operations. You'll see the "Delete Selected" button there. Keep in mind that this secondary top bar doesn't have to move the project list, it could be overlayed on top, since it's temporary and you'll only ever see it when you want to delete multiple projects

> I would prefer an actual "Open in file browser" action

Tbh probably same when I think about this one, a clickable link takes away from the clickable area of the list entry. It's a good candidate for the sandwich menu

> The big padding is not friendly for small screens IMHO. My project manager window does not open that big

Keep in mind that it may look bigger than it is in reality if you scroll this page and compare before/afters because we made the window smaller by removing the sidebar and you basically see it zoomed in now. I'm personally using a laptop as a primary computer, whenever I design something I think of a small screen first (which is also partially the reason I was interested in making a custom theme to reduce vertical spaces in lists and dock gaps in Godot). So personally I'm fairly confident that this design works on a small screen

P.S. If we value space we definitely don't want a sidebar for actions (you lose much more space compared to the version that adds 2 small icons per list item):

<img width="933" alt="image" src="https://github.com/godotengine/godot-proposals/assets/60579014/bc2e1328-3ae3-4620-9929-bf88153200f1">

**Reply by PumpkinPaul** [reply]
Looks great, I love what you've both done with this

**Reply by viveleltsi** [reply]
I like what you are doing here guys !  I have a small proposition : add an external link icon after or before the folder path to indicate it will open a new window (file explorer) like this one. I had to read the conversation to know that the link was clickable so the icon will help to know what the link will do and the fact this is a link and an action.


![image](https://github.com/godotengine/godot-proposals/assets/4752251/1178a022-1b55-4211-b9fa-fa7da9524bed)

**Reply by MewPurPur** [reply]
No strong feelings about the sidebar, I mean it's not too useful, but it'll have to be replaced by a right-mouse-button popup to still allow actions on multiple projects.

This means, this can't be done with buttons that can be clicked once to open a project (and it wasn't fitting anyway, it's clearly more like an item list or a filesystem)

I like the current theming better, this feels... Annoyingly modern? My main issue is with the top bar's massive spacing, especially in the current revision, and the random colored button. I prefer everything about the current top bar more than the reworked one.

**Reply by MewPurPur** [reply]
@viveleltsi The current icon is fine IMO. The issue with your proposition is that this kind of icon suggests opening a website link.

There are better ways to make it more suggestive, like a hover indicator.

**Reply by viveleltsi** [reply]
I agree with you but I also have seen this icon for opening an external something which is a local action and not on internet. But you are right.

**Reply by MewPurPur** [reply]
External applications generally. Which your browser is. But the filesystem isn't quite that.

**Reply by passivestar** [reply]
> My main issue is with the top bar's massive spacing

that can be reduced, especially the space under the topbar to make the list stick more to it visually. @mgdawber will have to figure out the spacing that works best in practice as he's coding this, it's an iterative process

**Reply by davthedev** [reply]
As a reference,  on Godot 4.2.1 stable, I can view 8 list items + a bit of the next one with the window opening at a height of 650px.
This is the very metric I care about; being able to view as many projects at a glance without scrolling. Lateral spacing is a bit less dramatic because you would mostly get shortened names that still have some space to expand.

The project list window is likely to be used in non-editing contexts as a frontend to run projects, such as just playing with demos and templates, or trying to run a game on an embedded machine with a smaller screen. Thinking of the Raspberry Pi and its official 800x480 display, as some users have attempted builds on this platform. Even if the main editor window is more complex, this introductory window with less content can be adapted to run in smaller screens for this usage as a game launcher.

Here is an attempt to match 4.2.1 current dialog with the mockup, scaled at best effort to match by the size of the project icons, when it opens at its default height of 650px on my 1920x1080 display. We lose a couple list items worth of screen real estate. Not that I am against making the thing more inviting, but we can strike a middle ground.

![image](https://github.com/godotengine/godot-proposals/assets/26961646/04e116b9-2824-4725-9f52-e85dfe4828c9)

An idea for that middle ground would be to propose a compact list mode, or a grid, or both.

The idea of a contextual action bar that overlays the list contents can be tricky in one way: the popup can get in your way in some situations, for example if you click to select the item at the very area where it is supposed to pop up. What about a bottom bar with the main actions, for example? (It defeats my point regarding use of vertical space, but I thought about the solution while writing)

![image](https://github.com/godotengine/godot-proposals/assets/26961646/8cc13904-74f8-448e-b5f6-78bad12c84bb)

**Reply by passivestar** [reply]
> We lose a couple list items worth of screen real estate

Sure, but you lose it to the top area that was introduced in 4.3 to replace the tabs as well. 4.2.1 isn't a fair comparison to this. I'm not against making it smaller, we can reduce the height of the top area, and we can reduce the space under the top bar, it'd be practically the same as 4.3 in terms of how much stuff fits vertically

> What about a bottom bar with the main actions, for example?

It's interesting but personally I'd prefer to make "Run" and "Edit" icon-only buttons and put them somewhere near the project title, could try making the Edit button larger to make it more touchscreen friendly. This will save you both vertical space and a click

**Reply by MewPurPur** [reply]
I still just like the old stuff better. I don't think "Create" needs any kind of particular highlight, people already create too many new projects as is. In all seriousness, I find that Create is not a common or a special action, and it's not more important than importing. 99 times out of a hundred, I go to Godot to edit my existing projects.

The sidebar takes space without being necessary, yes, but I don't find its removal to be an improvement either. The sidebar is great for seeing what actions are available outright. An action button is a compromise used to simplify the UI by reducing the amount of information frontloaded on the user, and also freeing up space. The project manager UI is very simple and it's not struggling to fit in the space, so this effort looks misguided. Also, I think the sidebar makes for a better default aspect ratio of the project list itself. Moreover, all these buttons are dimmed when you first start out, so it's not even that distracting, your attention immediately goes to the top options anyway.

So yeah, I personally don't like any of these attempts, I don't think they go in the right direction at all.

**Reply by EmrysMyrddin** [reply]
I pretty much disagree with you on this.

As a pretty new Godot user, I can say that the first impression when opening Godot and seeing this project manager is not very good. It feels overwhelming to have all those buttons and things that you don't even actually need.

Most of the time, you just go here to either create a project or open an existing one.

As stated by others, a lot of clickable items are not obviously clickable.

And as final point, even if I agree that it should not take over usability, the proposed mock-ups are more aligned with what is expected from a modern and professional look. It looks really great :-)

**Reply by mgdawber** [reply]
Thanks for all the feedback everyone, they're all valid concerns - much appreciated!

I have tried to implement fixes based off of everyone's main concerns since the last iteration.

As a summary of what I could see from the discussion:

- Padding is too wide, takes up too much screen estate.
- Preference for the old 'open/view directory' button. The blue with underline could be considered a hyperlink.
- Batch operations are missing. Running multiple instances for multiplayer and batch deletion are common enough to warrant this.

There is also valid concern that the design may stray too far from Godot's current design feel - I've made a few smaller adjustments to try and bring it back in line.

For this version, I've played around with the idea from @davthedev with the bottom bar as a compromise to losing the side bar.

How does everyone feel about this one - thoughts, likes, dislikes?

If the bottom bar doesn't get good reception then I'll also explore @passivestar's idea with the single edit/run icon near the project title!

![Version - 1](https://github.com/godotengine/godot-proposals/assets/61522828/7f646c65-4fff-453e-bf85-fa8ec9701f2b)

**Reply by davthedev** [reply]
Wow, seeing the mockups evolve helps generating new ideas!

I see two types of presentations that may serve different purposes and have their own audience. Both have valid reasons to exist. I think we can make both exist next to each other.

- The "newcomer" mode with the extra padding, gives the "wow" effect and the enjoyable first impression to casual users, that is enabled by default on a fresh "install"
- The "expert" mode that optimizes screen real estate with reduced margins. Would be enabled using a system-wide setting by the savvy user once the novelty has worn off and productivity becomes the priority. Can also automatically switch to it as a fallback in case the"newcomer" presentation cannot fit the current monitor resolution on a small screen.

**Reply by passivestar** [reply]
If being able to fit a lot of projects on screen is a big concern you can just add a compact list mode, not sure you need a system-wide setting for something this simple. Most OS'es file managers have this

![listmode](https://github.com/godotengine/godot-proposals/assets/60579014/8ead32df-8b26-47cf-89f4-94d2be153a4f)

**Reply by peterhoglund** [reply]
I'm so happy you changed the logo to the established white instead of the outlined thing proposed before. Here are my thoughts:

- I like the Edit and Run buttons. I have no strong opinion on whether they are at the bottom or shown on hover, but I think there needs to be clear actions for the most common use cases here. If they are at the bottom they should probably stand out a bit better, I didn't see them at first.
- In the latest image there are two Remove buttons, one at the bottom and one in each list element. Remove could easily be moved to the overflow menu. It is not that common that you remove projects. I think 95% of what users do here is create a new project or edit it. 
- The batch remove seems like a power user feature that doesn't have to be present at all times. The Remove button at the bottom and the "2 Projects selected" should only be visible when the user has selected more than 1 project, since 99% of the time only 1 project will be selected here.
- I'm a padding kind of guy :) Not a fan of super compact UIs, so I think the tags and the buttons can have some more padding to make them easier to read, but if I've understood it right you are proposing a "normal" and a "compact" version?

Awesome work!

**Reply by MewPurPur** [reply]
The "Favorite" icon probably shouldn't be yellow. Marking things as favorite is something that exists in a lot of places in Godot, and should use consistent theming. Making the star yellow would be distracting if it spills into the editor, where icons are normally white.

**Reply by PumpkinPaul** [reply]
this one is my favourite so far I think - I'm firmly in the camp of make Godot look cleaner / better and not in the make it look like the existing stuff :-)
If I could offer up one suggestion, can we make the icon white like the Godot text (not a dirty white like it is there) :-)

**Reply by mgdawber** [reply]
That compact view looks very nice!

Here's another version including the points mentioned above.

I've tried a hollow favorite icon here as well, not sure whether I prefer it from what it is now, but good to hear thoughts on it anyway.

![ProjectManagerScreen](https://github.com/godotengine/godot-proposals/assets/61522828/f15d1375-00ef-4715-b2a1-4f7a8f1049d6)

**Reply by EmrysMyrddin** [reply]
I'm not that much a fan of the bottom bar. And not sure why we need it ? Is it that common to make bulk actions ? What is the use case of running multiple projects at once ?

For deletion, it is understandable, but I feel that it takes a lot of place and have a lot of traction for something most users will probably not use often if at all.

I mean, if you take OS file explorer as an example, there isn't an "action bar" with all common actions you can do in bulk mode. You select the files and the right click on it to do your action. In macOS, we have a "burger button" in the top bar which opens the same menu than the right click.

I think that the compact view is a good addition when you start to have a lot of projects.

I personally prefer the top bar version of @passivestar with a bit more padding. It doesn't eat up a lot of place but makes the bar feeling more "solid" and visible.

**Reply by mgdawber** [reply]
I'm in agreement, I really like the way @passivestar's version looks 🚀 

It would be useful to know how common of a task opening multiple projects at once is - the only thing that comes to mind is server/client architecture which is served in different projects? I can't imagine it would be too common, but then again I'm quite new to game programming so insight from people more experienced would be useful.

Here's another version of what the bar could look like in a more compact form, _sorry for the spam!_ 

![MacBook Air](https://github.com/godotengine/godot-proposals/assets/61522828/852b7b2c-9b12-4ea9-88de-687038e3892d)

**Reply by passivestar** [reply]
> if you take OS file explorer as an example, there isn't an "action bar" with all common actions you can do in bulk mode. You select the files and the right click on it to do your action

Yep, or a long press on a touch screen

It doesn't take much to make an app that's easy to use when there's already an established interaction language that all of software has been using for decades. If you use that, the UI gets out of the way and users can focus on making games, it's as simple as that

**Reply by KoBeWi** [reply]
Some notes from me:
- Personally I run the projects from project manager relatively often, when the project is an app and I want to start it, or when I want to test a project in another version of Godot without editing it. In contrast, I almost never delete projects. I think having Run button available  on project entry would be useful, more than the Delete button.
- The main reason why tags use a dialog (other than it was easier to implement :P) is that the dialog shows **all** tags used by your project, so you can easily assign tags that you use in other projects. Will the inline adding support it?
- For the record, full color tags was their initial design, but it was changed to the current one after feedback (https://github.com/godotengine/godot/pull/75047)

**Reply by passivestar** [reply]
> The main reason why tags use a dialog (other than it was easier to implement :P) is that the dialog shows all tags used by your project, so you can easily assign tags that you use in other projects. Will the inline adding support it?

Inline editors usually use autocomplete to show all tags. They could also be sorted, most commonly used on top. Realistically if I was @mgdawber I would probably keep the Manage Tags thing in the right click / sandwich menu for now, and add the inline editing separately at a later point, because it will probably be a lot of work

**Reply by EmrysMyrddin** [reply]
@mgdawber Ah, this version of the bottom bar is really better :-) Perhaps we can make it appear only when at least 2 projects are selected ? Or just 1 if we really want to show that there is the possibility to have bulk actions. I've seen this pattern in multiple Apps :-)

I really think that what a new user would expect from this screen is double clicking on a line opens it in editor. So would probably ok to add a "run" button in place of the "remove" button that was present in previous mockups ?

**Reply by EmrysMyrddin** [reply]
As a note, I would probably not make the `Run` button the accentuated button. `Edit` is probably a better feat for it, since it's the main purpose of this software.

And perhaps the `Delete` button should show some danger, like with a red-ish color ?

**Comment by Ledracarnos** [comment]
I find it frustrating to add tags by opening an additional window, I have no idea what to do next, just a "+" button.
![uxgodotfeedb](https://github.com/godotengine/godot-proposals/assets/79541759/a17feae5-4a58-451f-a6e2-d7f87e3aecbe)

**Reply by atirutw** [reply]
The `+` button could be moved to the right for a more consistent position

**Reply by Ledracarnos** [reply]
I don't know, I'll put the result here just in case.
![uxgodotfeedb2](https://github.com/godotengine/godot-proposals/assets/79541759/199d57a2-f3bd-456c-9283-28e74ad0520c)

**Reply by mgdawber** [reply]
Nice, definitely think it's something worth considering - I'm working on a branch to test the feasibility of the overall design and think this might be better to look at separately so I'll leave it out for now, but I had an explore of what it potentially could look like as a simple dropdown anyway! 

![image 2](https://github.com/godotengine/godot-proposals/assets/61522828/cdde806d-35af-4608-b4c0-6554b9de7990)

_The scale is a bit off, but the idea is there at least_

**Comment by Ledracarnos** [comment]
I think the new design is better, with the star button more discreet. I think it's a good idea to hide the "remove" button.
However, I'm not convinced by the run button on the right, I feel like I'm listening to music on spootify.

What's more, I don't see the "edit" button. I think the user has the right to ignore the fact that he can double-click on the project. 

If an execution button is to appear on the project, it must, in my opinion, stand out from the button with the three little dots.
![image](https://github.com/godotengine/godot-proposals/assets/79541759/5006c0b4-722f-4967-93c3-9cb835780a56)
Also, have you thought of making the edit/run button appear only when the mouse hovers over the project?

**Reply by mgdawber** [reply]
Hah, it does look like music tracks :p 

My worry was that hiding the run button and only appearing on hover/selected would cause the tags to jump, unless we gave the run button static space, but that might look a little odd. Though now that you mention spotify and only showing play on hover.. could always do what they do.. 😆 

![MacBook Air - 1 (1)](https://github.com/godotengine/godot-proposals/assets/61522828/ca7ca4aa-b906-484b-abae-eedc6381658c)

**Reply by Ledracarnos** [reply]
I'm not convinced ^^''' 
I think the contextual bottom bar is a very good idea, maybe even sufficient. But we need to gather more opinions. 

edit: Looking at all the feedback, I think a bottom bar that's always visible is a very good idea.

**Reply by davthedev** [reply]
I think the icons-appearing-on-hover-or-select does not play well in a list of contents where you can select items, having the selection click not doing any actual action other than selecting the items.

This pattern is really a web thing, commonly associated with lists that have no concept of "selecting items". Only thing you can do is click on the individual actions proposed on each row.

Something also coming in mind: I get it that this is a mockup, but make sure to give a different look to buttons and text input fields at the end of the day. Those do not behave the same; thus those should not resembling each other with the same background and border style - different text styling is not enough because the text field can be empty, or hint text may be confused with a disabled button.

**Comment by mgdawber** [comment]
Here's a version which is a bit less minimal, akin more towards a file explorer.

I've tried to collect actions into groups:
- The top left bar is for adding projects.
- The top right of the bar is for project actions. The text is to give a visual cue that these can be batched.
- And the list container bar for sort/filter/list view.

I've resized the mockup to a more appropriate size (1280x823) so we can see what the padding looks like, and populated the list fully.

![MacBook Air - 2 (1)](https://github.com/godotengine/godot-proposals/assets/61522828/0b7bc6af-2185-42a7-ace2-b34d46069d07)
![MacBook Air - 3 (1)](https://github.com/godotengine/godot-proposals/assets/61522828/2a6f8910-05b0-49aa-a50c-e64361e5fb4d)

**Reply by passivestar** [reply]
this layout is interesting. are you sure the Run button is as important as the Edit button and needs color?

**Reply by Ledracarnos** [reply]
It's always a pleasure to see the evolution, it's really better, I don't know what to say. 

Among the ideas proposed, this one was interesting, it makes it clear that the options are clickable. 
![image](https://github.com/godotengine/godot-proposals/assets/79541759/086deb54-1c16-41bb-963f-582789bbb3af)
Maybe these outlines could appear when the mouse suvoles them.

As for the run and edit buttons, I really like them, but it's true that they can be confusing. But is it the colors or the names?
Shouldn't the "run" button be called "play"?

Edit : Perhaps a clearer hierarchy is needed. 
![image](https://github.com/godotengine/godot-proposals/assets/79541759/cec4ad55-2902-41db-8d9c-b1b7656198dd)
Sorry for the drafts ^^

**Reply by mgdawber** [reply]
Yeah it was probably a bit too much - I've updated and switched them round :)

**Reply by davthedev** [reply]
Nice demo for a compact mode! When having dozens of projects in the list, this can help finding our stuff quicker!

Perhaps make the selected item have a different color than the frame so that it is not fused with it, for example a shade of blue that stands out more. Or the first item may appear to be fused with the toolbar when selected.

> Maybe these outlines could appear when the mouse suvoles them.

Yep, each actionable element should have immediate action feedback with hover and pressed / "button down" appearances. But, be it an outline, a bg color change, there are multiple creative possibilities to try out.

> a more appropriate size (1280x823)

Keep in mind that the window should work well on small screens too, therefore any resolution should be appropriate in theory. Something too much overlooked in this era.
